nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -76,13 +78,90 @@ impl InstanceAndActiveVmm {
+        use crate::db::model::InstanceState;
+        use crate::db::model::VmmState;
+
+        let instance_state = instance.runtime_state.nexus_state;
+        let vmm_state = active_vmm.map(|vmm| vmm.runtime.state);
+
+        // We want to only report that an instance is `Stopped` when a new
+        // `instance-start` saga is able to proceed. That means that:
+        match (instance_state, vmm_state) {
+            // - If there's an active migration ID for the instance, *always*
+            //   treat its state as "migration" regardless of the VMM's state.
+            //
+            //   This avoids an issue where an instance whose previous active
+            //   VMM has been destroyed as a result of a successful migration
+            //   out will appear to be "stopping" for the time between when that
+            //   VMM was reported destroyed and when the instance record was
+            //   updated to reflect the migration's completion.
+            //
+            //   Instead, we'll continue to report the instance's state as
+            //   "migrating" until an instance-update saga has resolved the
+            //   outcome of the migration, since only the instance-update saga
+            //   If the instance actually *has* stopped or failed before a
+            //   successful migration out, this is fine, because an
+            //   instance-update saga will come along and remove the active VMM
+            //   and migration IDs.
+            //
+            (InstanceState::Vmm, Some(_))
+                if instance.runtime_state.migration_id.is_some() =>
+            {
+                external::InstanceState::Migrating
+            }
+            // - An instance with a "stopped" or "destroyed" VMM needs to be
+            //   recast as a "stopping" instance, as the virtual provisioning
+            //   resources for that instance have not been deallocated until the
+            //   active VMM ID has been unlinked by an update saga.
+            (
+                InstanceState::Vmm,
+                Some(
+                    VmmState::Stopped
+                    | VmmState::Destroyed
+                    | VmmState::SagaUnwound,
+                ),
+            ) => external::InstanceState::Stopping,
+            // - An instance with no VMM is always "stopped" (as long as it's
+            //   not "starting" etc.)
+            (InstanceState::NoVmm, _vmm_state) => {
+                debug_assert_eq!(_vmm_state, None);
+                external::InstanceState::Stopped
+            }
+            // If there's a VMM state, and none of the above rules apply, use
+            // that.
+            (_instance_state, Some(vmm_state)) => {
+                debug_assert_eq!(_instance_state, InstanceState::Vmm);
+                vmm_state.into()
+            }
+            // If there's no VMM state, use the instance's state.
+            (instance_state, None) => instance_state.into(),
```

Comment:
This bit is actually worthy of some consideration: because an instance's external-API effective state indicates which user/API-initiated state transitions it can undergo, and we determine that state based on both the instance and VMM states, we have to do a bit of munging to make sure it is consistent with the permitted operations. 

Some similar logic used to live in some of the different instance API endpoints, but now all of it lives here, and we've added some new rules based on the changes in this PR.